### PR TITLE
rpc: implement MPT-related RPC (2.x)

### PR DESCRIPTION
**Pull request by fyrchik**
Implement missing RPC calls and P2P handlers.

### CODE REVIEW

File: pkg/network/server.go [hunk 1]
```diff
@@ -509,7 +514,10 @@ func (s *Server) handleGetDataCmd(p Peer, inv *payload.Inventory) error {
 				msg = s.MkMsg(CMDBlock, b)
 			}
 		case payload.StateRootType:
-			return nil // do nothing
+			r := s.stateCache.Get(hash)
```
Comment (roman-khimov) [inline pkg/network/server.go:517]: But what's the point of caching it, really? We have all the states in the DB anyway and can answer this call using that data.

File: pkg/network/server.go [hunk 2]
```diff
@@ -613,12 +621,17 @@ func (s *Server) handleGetRootsCmd(p Peer, gr *payload.GetStateRoots) error {
...
+		_ = s.chain.AddStateRoot(&rs.Roots[i])
+	}
+	return nil
 }
 
 // handleStateRootCmd processees `stateroot` request.
 func (s *Server) handleStateRootCmd(r *state.MPTRoot) error {
-	return nil // TODO
+	// we ignore error, because there is nothing wrong if we already have this state root
+	_ = s.chain.AddStateRoot(r)
```
Comment (roman-khimov) [inline pkg/network/server.go:633]: It actually should be rebroadcasted if it's something new for us.

File: pkg/rpc/server/server_test.go [hunk 3]
```diff
@@ -214,6 +214,28 @@ var rpcTestCases = map[string][]rpcTestCase{
 			},
 		},
 	},
+	"getproof": {
```
[the remark this answers is not in the document]
Comment (fyrchik) [reply pkg/rpc/server/server_test.go:218]: It is below together with verify.

File: pkg/network/server.go (hunk 1 above)
Comment (fyrchik) [reply pkg/network/server.go:517]: Well, there are 3 types of hashes: state root itself, hash of `StateRoot` struct and hash of the block. I don't think it is worth to have 1 and 2 in the blockchain itself. `StateRoot` from arbitrary blocks can be queried together via `StateRoots` payload and `Inventory` is usually used for new payloads.

Comment (fyrchik) [reply pkg/network/server.go:517]: Actually 4, there is also hash of the P2P payload. And it is completely unrelated to anything `Blockchain` should deal with IMO.

File: pkg/rpc/response/result/mpt.go [hunk 4]
```diff
@@ -1,7 +1,74 @@
...
+// MarshalJSON implements json.Marshaler.
+func (p *ProofWithKey) MarshalJSON() ([]byte, error) {
+	return []byte(`"` + p.EncodeString() + `"`), nil
+}
+
+// EncodeString transforms p into a hex-encoded string.
+func (p *ProofWithKey) EncodeString() string {
+	w := io.NewBufBinWriter()
+	w.BinWriter.WriteVarBytes(p.Key)
+	w.BinWriter.WriteVarUint(uint64(len(p.Proof)))
```
[the remark this answers is not in the document]
Comment (fyrchik) [reply pkg/rpc/response/result/mpt.go:45]: Fixed.

File: pkg/rpc/server/server.go [hunk 5]
```diff
@@ -687,6 +688,28 @@ func (s *Server) getDecimals(h util.Uint160, cache map[util.Uint160]int64) (int6
...
+	}
+	var rt *state.MPTRootState
+	var h util.Uint256
+	height, err := p.GetInt()
+	if err == nil {
+		rt, err = s.chain.GetStateRoot(uint32(height))
+	} else if h, err = p.GetUint256(); err == nil {
+		hdr, err := s.chain.GetHeader(h)
+		if err == nil {
+			rt, err = s.chain.GetStateRoot(hdr.Index)
```
[the remark this answers is not in the document]
Comment (fyrchik) [reply pkg/rpc/server/server.go:768]: This API is not a part of this PR.

File: pkg/rpc/response/result/mpt.go [hunk 6]
```diff
@@ -72,3 +80,36 @@ func (p *ProofWithKey) UnmarshalJSON(data []byte) error {
...
+		return []byte(`"invalid"`), nil
+	}
+	return []byte(`{"value":"` + hex.EncodeToString(p.Value) + `"}`), nil
+}
+
+// UnmarshalJSON implements json.Unmarshaler.
+func (p *VerifyProof) UnmarshalJSON(data []byte) error {
+	var m map[string]string
+	if err := json.Unmarshal(data, &m); err != nil {
+		if bytes.Equal(data, []byte(`"invalid"`)) {
```
[the remark this answers is not in the document]
Comment (fyrchik) [reply pkg/rpc/response/result/mpt.go:96]: Fixed.

File: pkg/network/server.go (hunk 2 above)
Comment (fyrchik) [reply pkg/network/server.go:633]: `StateRoot` is a response for `GetData` command, we have already rebroadcasted it's hash.

File: pkg/rpc/request/param.go [hunk 7]
```diff
@@ -79,6 +79,21 @@ func (p *Param) GetString() (string, error) {
...
 }
 
+// GetBoolean returns boolean value of the parameter.
+func (p *Param) GetBoolean() bool {
+	if p == nil {
+		return false
+	}
+	switch p.Type {
+	case NumberT:
+		return p.Value != 0
```
[the remark this answers is not in the document]
Comment (fyrchik) [reply pkg/rpc/request/param.go:89]: JSON spec does not have a notion of `NaN`, may be this check is a result of a silent error in C# unmarshaler. `Go` unmarshals null into `float64(0)`. Made a shortcut for this case.

File: pkg/network/server.go (hunk 2 above)
Comment (roman-khimov) [reply pkg/network/server.go:633]: > StateRoot is a response for GetData command

Not always, see neo-project/neo#1662, but at the same time it's very similar to a block or tx.

> we have already rebroadcasted it's hash.

Where? It should be handled very similar to block (technically even with something similar to `blockQueue` for roots corresponding to not-yet-accepted blocks.

File: pkg/rpc/server/server_test.go (hunk 3 above)
[the remark this answers is not in the document]
Comment (roman-khimov) [reply pkg/rpc/server/server_test.go:218]: ok

File: pkg/network/server.go (hunk 2 above)
Comment (fyrchik) [reply pkg/network/server.go]: Fixed.

**Pull request merged**